ouroboros-network/src/Ouroboros/Network/PeerSelection/PeerStateActions.hs

```diff
@@ -757,15 +783,32 @@ withPeerStateActions timeout
             pchPeerState,
             pchAppHandles } = do
       -- quiesce warm peer protocols and set hot ones in 'Continue' mode.
-      atomically $ do
-        writeTVar pchPeerState PromotingToHot
-        writeTVar (getControlVar TokHot pchAppHandles) Continue
-        writeTVar (getControlVar TokWarm pchAppHandles) Quiesce
+      continue <- atomically $ do
+        -- if the peer is cold we can't activate it.
+        notCold <- updateUnlessCold pchPeerState PromotingToHot
+        when notCold $ do
+          writeTVar (getControlVar TokHot pchAppHandles) Continue
+          writeTVar (getControlVar TokWarm pchAppHandles) Quiesce
+        return notCold
+      when (not continue) $ do
+        traceWith spsTracer (PeerStatusChangeFailure
+                             (WarmToHot pchConnectionId)
+                             ActiveCold)
+        throwIO $ ColdActivationException pchConnectionId
```

Comment:
In `Governor.Activate.jobPromoteWarmPeer` there are these two `TODO` :

>     handler e =
>       --TODO: decide what happens if promotion fails, do we stay warm or go to
>       -- cold? Will this be reported asynchronously via the state monitoring?

>     job = do
>       --TODO: decide if we should do timeouts here or if we should make that
>       -- the responsibility of activatePeerConnection

We should update the first one, and if we are on this we should update the other one too.  Could you do that even though it's only slightly out of scope of this PR?

Reply:
I will update address the first TODO by setting an activation timer (testing showed that this is needed). 
I've removed the second comment and replaced it with a NB in `activatePeerConnection` . We should handle the timeout there so that it matches `establishPeerConnection` and `deactivatePeerConnection`.